File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/bootstrap/PlaylistPlaybackBootstrapper.java
package com.lasthopesoftware.bluewater.client.playback.engine.bootstrap;

import com.lasthopesoftware.bluewater.client.playback.engine.ActivePlayer;
import com.lasthopesoftware.bluewater.client.playback.engine.IActivePlayer;
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlayableFileQueue;
import com.lasthopesoftware.bluewater.client.playback.playlist.playablefile.PlayableFilePlayer;
import com.lasthopesoftware.bluewater.client.playback.volume.PlaylistVolumeManager;

public final class PlaylistPlaybackBootstrapper implements StartAndClosePlayback {

	private final PlaylistVolumeManager volumeManagement;

	private PlayableFilePlayer playlistPlayer;
	private ActivePlayer activePlayer;

	public PlaylistPlaybackBootstrapper(PlaylistVolumeManager volumeManagement) {
		this.volumeManagement = volumeManagement;
	}

	@Override
	public IActivePlayer startPlayback(PreparedPlayableFileQueue preparedPlaybackQueue, final long filePosition) {
		close();

		playlistPlayer = new PlayableFilePlayer(preparedPlaybackQueue, filePosition);
		activePlayer = new ActivePlayer(playlistPlayer, volumeManagement);

		return activePlayer;
	}

	@Override
	public void close() {
		if (activePlayer != null) activePlayer.close();
		if (playlistPlayer != null)	playlistPlayer.close();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/bootstrap/PlaylistPlaybackBootstrapper.kt
package com.lasthopesoftware.bluewater.client.playback.engine.bootstrap

import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlayableFileQueue
import com.lasthopesoftware.bluewater.client.playback.playlist.ManagePlaylistPlayback
import com.lasthopesoftware.bluewater.client.playback.playlist.PlaylistPlayer
import com.lasthopesoftware.bluewater.client.playback.volume.PlaylistVolumeManager
import org.joda.time.Duration
import java.io.Closeable

class PlaylistPlaybackBootstrapper(private val volumeManagement: PlaylistVolumeManager) : IStartPlayback, Closeable {

	@Volatile
	private var playlistPlayer: PlaylistPlayer? = null

	@Synchronized
	override fun startPlayback(preparedPlaybackQueue: PreparedPlayableFileQueue, filePosition: Duration): ManagePlaylistPlayback {
		playlistPlayer?.close()
		val newPlayer = PlaylistPlayer(preparedPlaybackQueue, filePosition)
		volumeManagement.managePlayer(newPlayer)
		playlistPlayer = newPlayer
		newPlayer.prepare()
		return newPlayer
	}

	override fun close() {
		playlistPlayer?.close()
	}
}
